### PR TITLE
Twinning Propagation model for FCC material

### PR DESCRIPTION
**Pull request by sapitts**
Refs #18846 

Implements the twinning propagation model from Kalidindi IJP (2001) for face-centered cubic materials. This twinning propagation model is a foundational reference for other constitutive models for twin propagation, including HCP materials.

## Design
Implements and tests the modification of the plastic velocity gradient calculation to include the impact of the total twin volume fraction on the contribution from dislocation glide, following the form proposed by Kalidindi JMPS (1998) and Kalidindi IJP (2001).

### CODE REVIEW

File: modules/tensor_mechanics/include/materials/crystal_plasticity/CrystalPlasticityTwinningKalidindiUpdate.h [hunk 1]
```diff
@@ -0,0 +1,131 @@
...
+//*
+//* Licensed under LGPL 2.1, please see LICENSE for details
+//* https://www.gnu.org/licenses/lgpl-2.1.html
+
+#pragma once
+
+#include "CrystalPlasticityStressUpdateBase.h"
+
+class CrystalPlasticityTwinningKalidindiUpdate;
+
```
Comment (dschwen) [inline modules/tensor_mechanics/include/materials/crystal_plasticity/CrystalPlasticityTwinningKalidindiUpdate.h:15]:
```suggestion
```

File: modules/tensor_mechanics/src/materials/crystal_plasticity/CrystalPlasticityKalidindiUpdate.C [hunk 2]
```diff
@@ -37,7 +42,13 @@ CrystalPlasticityKalidindiUpdate::CrystalPlasticityKalidindiUpdate(
...
     _ao(getParam<Real>("ao")),
     _xm(getParam<Real>("xm")),
-    _gss_initial(getParam<Real>("gss_initial"))
+    _gss_initial(getParam<Real>("gss_initial")),
+
+    _include_twinning_in_Lp(parameters.isParamSetByUser("total_twin_volume_fraction")),
+    _twin_volume_fraction_total(
+        _include_twinning_in_Lp ? &getMaterialPropertyByName<Real>(
+                                      getParam<MaterialPropertyName>("total_twin_volume_fraction"))
+                                : nullptr)
```
Comment (dschwen) [inline modules/tensor_mechanics/src/materials/crystal_plasticity/CrystalPlasticityKalidindiUpdate.C:51]: why not just 
```suggestion
    _twin_volume_fraction_total(
        _include_twinning_in_Lp ? &getMaterialProperty<Real>("total_twin_volume_fraction"))
                                : nullptr)
```

File: modules/tensor_mechanics/include/materials/crystal_plasticity/CrystalPlasticityTwinningKalidindiUpdate.h (hunk 1 above)
Comment (sapitts) [reply modules/tensor_mechanics/include/materials/crystal_plasticity/CrystalPlasticityTwinningKalidindiUpdate.h:15]: done, thanks!

File: modules/tensor_mechanics/src/materials/crystal_plasticity/CrystalPlasticityKalidindiUpdate.C (hunk 2 above)
Comment (sapitts) [reply modules/tensor_mechanics/src/materials/crystal_plasticity/CrystalPlasticityKalidindiUpdate.C:51]: I thought we'd had a discussion in standup about how `getMaterialPropertyByName` was the better option; is that not the case? If so, I will happily make the change and update my notes

Comment (dschwen) [reply modules/tensor_mechanics/src/materials/crystal_plasticity/CrystalPlasticityKalidindiUpdate.C:51]: I don't recall such a discussion. But we can talk about it tomorrow.

Comment (sapitts) [reply modules/tensor_mechanics/src/materials/crystal_plasticity/CrystalPlasticityKalidindiUpdate.C:51]: correction made and notes updated

File: modules/tensor_mechanics/src/materials/crystal_plasticity/CrystalPlasticityKalidindiUpdate.C [hunk 3]
```diff
@@ -24,6 +24,11 @@ CrystalPlasticityKalidindiUpdate::validParams()
   params.addParam<Real>("xm", 0.1, "exponent for slip rate");
   params.addParam<Real>("gss_initial", 60.8, "initial lattice friction strength of the material");
 
+  params.addParam<MaterialPropertyName>(
+      "total_twin_volume_fraction",
+      "Name of the material property of the total twin volume "
+      "fraction, only if twinning is considered in the simulation");
+
```
Comment (dschwen) [inline modules/tensor_mechanics/src/materials/crystal_plasticity/CrystalPlasticityKalidindiUpdate.C:31]:
```suggestion
      "total_twin_volume_fraction",
      "Total twin volume fraction, if twinning is considered in the simulation");

```

I'd prefer that over explicitly stating "Name of the material property", which is implied by the type of the parameter, and also a bit misleading, as the user can supply a number value directly.

File: modules/tensor_mechanics/src/materials/crystal_plasticity/CrystalPlasticityKalidindiUpdate.C [hunk 4]
```diff
@@ -37,7 +42,12 @@ CrystalPlasticityKalidindiUpdate::CrystalPlasticityKalidindiUpdate(
     _gss_a(getParam<Real>("gss_a")),
     _ao(getParam<Real>("ao")),
     _xm(getParam<Real>("xm")),
-    _gss_initial(getParam<Real>("gss_initial"))
+    _gss_initial(getParam<Real>("gss_initial")),
+
+    _include_twinning_in_Lp(parameters.isParamSetByUser("total_twin_volume_fraction")),
```
Comment (dschwen) [inline modules/tensor_mechanics/src/materials/crystal_plasticity/CrystalPlasticityKalidindiUpdate.C:47]:
```suggestion
    _include_twinning_in_Lp(parameters.isParamValid("total_twin_volume_fraction")),
```

File: modules/tensor_mechanics/src/materials/crystal_plasticity/CrystalPlasticityTwinningKalidindiUpdate.C [hunk 5]
```diff
@@ -0,0 +1,339 @@
...
+        getParam<Real>("non_coplanar_coefficient_twin_hardening")),
+    _coplanar_coefficient_twin_hardening(getParam<Real>("coplanar_coefficient_twin_hardening")),
+    _noncoplanar_exponent(getParam<Real>("non_coplanar_twin_hardening_exponent")),
+    _limit_twin_volume_fraction(getParam<Real>("upper_limit_twin_volume_fraction"))
+{
+  // resize local caching vectors used for substepping
+  _previous_substep_twin_resistance.resize(_number_slip_systems);
+  _previous_substep_twin_volume_fraction.resize(_number_slip_systems);
+  _twin_resistance_before_update.resize(_number_slip_systems);
+  _twin_volume_fraction_before_update.resize(_number_slip_systems);
```
Comment (dschwen) [inline modules/tensor_mechanics/src/materials/crystal_plasticity/CrystalPlasticityTwinningKalidindiUpdate.C:103]: those should be initialized to that size in the initializer list

File: modules/tensor_mechanics/src/materials/crystal_plasticity/CrystalPlasticityTwinningKalidindiUpdate.C [hunk 6]
```diff
@@ -0,0 +1,339 @@
...
+
+  // Resize constitutive-model specific material properties
+  _twin_volume_fraction[_qp].resize(_number_slip_systems);
+  _twin_volume_fraction_increment[_qp].resize(_number_slip_systems);
+
+  // Set constitutive-model specific initial values from parameters
+  _total_twin_volume_fraction[_qp] = _initial_total_twin_volume_fraction;
+  const Real twin_volume_fraction_per_system =
+      _initial_total_twin_volume_fraction / _number_slip_systems;
+  for (unsigned int i = 0; i < _number_slip_systems; ++i)
```
Comment (dschwen) [inline modules/tensor_mechanics/src/materials/crystal_plasticity/CrystalPlasticityTwinningKalidindiUpdate.C:119]: The cool kids use
```suggestion
  for (auto i: make_range(_number_slip_systems))
```

these days (requires `#inlcude "libmesh/int_range.h"`)

**Review by dschwen — DISMISSED** [review]
Looks good to me, just some minor comments. No need for an AD version here?

File: modules/tensor_mechanics/src/materials/crystal_plasticity/CrystalPlasticityTwinningKalidindiUpdate.C (hunk 6 above)
Comment (sapitts) [reply modules/tensor_mechanics/src/materials/crystal_plasticity/CrystalPlasticityTwinningKalidindiUpdate.C:119]: well nothing about me is cool, so I guess that means I can ignore this suggestion ;-)

**Comment by sapitts** [comment]
No AD base class, so an AD version is out of scope for this PR. With 36-48 material properties at each qp for the simplest of crystal plasticity models, I'm not sure if AD makes sense for crystal plasticity...

File: modules/tensor_mechanics/src/materials/crystal_plasticity/CrystalPlasticityTwinningKalidindiUpdate.C (hunk 5 above)
Comment (sapitts) [reply modules/tensor_mechanics/src/materials/crystal_plasticity/CrystalPlasticityTwinningKalidindiUpdate.C]: I hope I made this change correctly, but if not please point me in the right direction

**Review by dschwen — APPROVED** [review]
Looks great now

**Pull request merged**